### PR TITLE
Add id to ref

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: pil_analyzer/src/json_exporter/mod.rs [hunk 1]
```diff
@@ -314,22 +314,27 @@ impl<'a> Exporter<'a> {
 
     fn polynomial_reference_to_json(
         &self,
-        PolynomialReference { name, index, next }: &PolynomialReference,
+        PolynomialReference {
+            name: _,
```
Comment (Schaeff) [inline pil_analyzer/src/json_exporter/mod.rs:318]: Is `_` required here?

File: pil_analyzer/src/lib.rs [hunk 2]
```diff
@@ -183,16 +184,40 @@ pub enum Expression {
...
-    // TODO would be better to use numeric IDs instead of names,
-    // but the IDs as they are overlap. Maybe we can change that.
+    /// Name of the polynomial - just for informational purposes.
+    /// Comparisons are based on polynomial ID.
     pub name: String,
+    /// Identifier for a polynomial reference.
+    /// Optional because it is filled in in a second stage of analysis.
+    /// TODO make this non-optional
+    pub poly_id: Option<(u64, PolynomialType)>,
     pub index: Option<u64>,
```
Comment (Schaeff) [inline pil_analyzer/src/lib.rs:196]: So this one becomes `None` after `poly_id` is filled? If so, would be good to document that here.

Comment (Schaeff) [reply pil_analyzer/src/lib.rs:196]: If it's always one of the other, would it make sense to have:

```
struct PolynomialReference<Id> {
   ..
   index: Id
}
```

And go from `PolynomialReference<u64>` to `PolynomialReference<(u64, PolynomialType)>` in the second stage of analysis? That would remove a lot of the runtime assumptions introduced in this PR (at the cost of making the AST generic over `Id` but maybe that's okay?)

File: pil_analyzer/src/display.rs [hunk 3]
```diff
@@ -211,12 +211,14 @@ namespace T(65536);
     col witness instr_dec_CNT;
     col witness instr_assert_zero;
     (T.instr_assert_zero * (T.XIsZero - 1)) = 0;
+    col witness X;
```
Comment (Schaeff) [inline pil_analyzer/src/display.rs:214]: Why did this change?

**Review by Schaeff — COMMENTED** [review]
I think we should at least consider making the AST generic on the `Id` to simplify the data structures here

File: pil_analyzer/src/display.rs (hunk 3 above)
Comment (chriseth) [reply pil_analyzer/src/display.rs:214]: Because we now check if referenced columns exist. This was referenced but didn't exist.

**Comment by chriseth** [comment]
Created https://github.com/chriseth/powdr/issues/232 about the poly_id being generic.

File: pil_analyzer/src/json_exporter/mod.rs (hunk 1 above)
Comment (chriseth) [reply pil_analyzer/src/json_exporter/mod.rs:318]: Yes, because it is not used.

File: pil_analyzer/src/lib.rs (hunk 2 above)
Comment (chriseth) [reply pil_analyzer/src/lib.rs:196]: But yes, I have considered making it generic, where the first-stage structure would have `poly_id: ()`, and the second-stage would be `poly_id: (u64, PolynomialType)`. Created issue #232  for that.

**Review by Schaeff — APPROVED** [review]

**Pull request merged**